doc/storagemond/storagemond-hld.md

```diff
@@ -0,0 +1,452 @@
+
+#### **2.4.3 Support for common implementations**
+
+Specific data, such as Filesystem Input/Output (FS IO) Reads/Writes, can be uniformly collected regardless of the storage disk type, as it is extracted from files generated by the Linux Kernel. To streamline the process of gathering this information, we propose the implementation of a new parent class `StorageCommon()`, from which classes such as SsdUtil, EmmcUtil, USBUtil, and NVMUtil would inherit in addition to `SsdBase` (to be renamed `StorageBase`). This proposed class will reside in the `src/sonic-platform-common/sonic_platform_base/sonic_ssd` directory, in `storage_common.py`. The `StorageCommon()` class will have the following functions:
+
+```
+def get_fs_io_reads(self):
+    """
+    Function to get the latest reads on each disk by parsing the /proc/diskstats file
+
+    Returns:
+        The total number of FSIO reads
+    
+    Args:
+        N/A
+    """
+
+def get_fs_io_writes(self):
+    """
+    Function to get the latest writes on each disk by parsing the /proc/diskstats file
+
+    Returns:
+        The total number of FSIO writes
+    
+    Args:
+        N/A
+    """
+```
+
+**Logic for StorageCommon() get_fs_io_reads and get_fs_io_writes functions:**
+
+These two functions, `get_fs_io_reads` and `get_fs_io_writes`, are designed to retrieve the latest disk reads and writes, respectively, by parsing the `/proc/diskstats` file. They utilize similar logic, differing only in the column index used to extract the relevant information.
+
+1. **Check for `psutil` Module**:
+   - The functions first check if the `psutil` module is available in the current environment by examining the `sys.modules` dictionary.
+
+2. **Use `psutil` Module (if available)**:
+   - If `psutil` is available:
+     - The functions retrieve disk I/O counters, specifying the disk for which to get the counters.
+     - They then get the read or write count for the specified disk using `read_count` or `write_count` respectively.
+
+3. **Fallback to Parsing Disk Stats File**:
+   - If `psutil` is not available:
+     - The functions open the `/proc/diskstats` file
+     - They read the contents of the file and iterate over each line.
+     - For each line, they check if the name of the storage disk is present.
+     - If the name of the storage disk is found in the line, they return the value at the appropriate zero-based index (3 for reads, 7 for writes).
+     - If no line contains the name of the storage disk, they save the respective values as 0.
+
+4. **Return to caller**:
+     - These values are then returned to the caller to subsequently be written to `STATE_DB`.
+     - These values are also used to determine the total number of procfs reads and writes as explained in the following section
+
+
+#### **2.4.4 Accounting for reboots and unintended powercycles**
```

Comment:
@assrinivasan please add a sudo code for the daemon accounting all scenarios.

Reply:
Done in latest commit